docs/design/proposals/collections.md

```diff
@@ -0,0 +1,604 @@
+- Custom aggregations across the collection
+
+### includeWhen with Collections
+
+The resource-level `includeWhen` determines whether any instances should be
+created at all. If `includeWhen` evaluates to false, the entire forEach is
+skipped and no resources are created.
+
+```yaml
+- id: backupJobs
+  includeWhen: ${size(spec.databases) > 0 && spec.backupEnabled}
+  forEach: ${spec.databases}
+  template:
+    kind: CronJob
+    # ...
+```
+
+This is different from filtering within forEach - it's an all-or-nothing
+decision for the entire collection.
+
+### excluding specific instances within a collection
+
+To exclude specific instances from being created within a collection, users can
+leverage CEL functions like `filter` or `map` directly in the `forEach`
+expression. For example i want to create N configMaps and exclude odd numbers:
+
+```yaml
+- id: configMaps
+  forEach: ${range(0, 10).filter(x, x % 2 == 0)} # Only even numbers
+  template:
+    kind: ConfigMap
+    metadata:
+      name: ${'config-' + string(each.item)} # config-0, config-2, config-4, ...
+```
+
+### simple schema enhancements
+
+The status section of the schema can now include CEL expressions that reference
+resource collections. This enables dynamic status computation based on the state
+of all resources created by forEach:
+
+```yaml
+apiVersion: kro.run/v1alpha1
+kind: ResourceGraphDefinition
+spec:
+  schema:
+    apiVersion: v1alpha1
+    kind: Application
+    spec:
+      workers: []string
+    status:
+      # Reference collections directly in status
+      totalWorkers: ${size(workerPods.items)}
+      runningWorkers: ${size(workerPods.items.filter(w, w.status.phase == 'Running'))}
+      allWorkersReady: ${workerPods.all(w, w.status.conditions.exists(c, c.type == 'Ready' && c.status == 'True'))}
+      workerNames: ${workerPods.items.map(w, w.metadata.name).join(',')}
+  resources:
+    - id: workerPods
+      forEach: ${spec.workers}
+      template:
+        kind: Pod
+        # ...
+```
+
+Status fields are continuously updated as the collection changes, providing
+real-time observability into the state of all resources managed by `forEach`.
+
+### resource references in CEL
+
+Colelction resourecs can be referenced by other resources using their "ID". For
+example, you have a collection of services and you want to create an ingress
+that routes to all of them:
+
+```yaml
+kind: ResourceGraphDefinition
+# ...
+spec:
+  resources:
+    - id: microservices
+      forEach: ${spec.services}
+      template:
+        kind: Service
+        # ...
+
+    - id: gateway
+      template:
+        kind: Ingress
+        # ...
+        spec:
+          rules:
+            - host: ${spec.domain}
+              http:
+                # collection resource can be referenced by its ID
+                paths:
+                  | # kro will do it's magic making sure to replace this with the proper type.
+                  ${microservices.map(svc, {
```

Comment:
Will this cel expression re-evaluate on any change to any service? Maybe somewhat challenging. Would that lead users to optimize performance with something like `forEach: ${spec.services.map(s, s.status.loadbalancer.ingress.hostname[0])`. This would resolve the services in isolation on line 325 before doing the more expensive evaluation on line 341.

Reply:
>Will this cel expression re-evaluate on any change to any service? 

Correct to change to any service or every "resyncPeriod" duration.

>This would resolve the services in isolation on line 325 before doing the more expensive evaluation on line 341.

Are you suggesting users could reference "static" values directly from the instance spec rather than from the dynamic service resources?

Reply:
Discussed offline -- I think this is essentially the equivalent of a compiler optimization. We can essentially look at subtrees of the DAG, detect changes at specific nodes, and only re-evaluate children within the DAG. For large DAGs, this could result in significant performance improvements.